cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/PluginService.java

```diff
@@ -101,6 +111,60 @@ private ArtifactDescriptor pickParentArtifact(ArtifactDetail artifactDetail, Id.
     // return the first one from the artifact details list.
     return artifactDetails.get(0).getDescriptor();
   }
+  /**
+   * A RemovalListener for closing classloader and plugin instantiators.
+   */
+  private static final class InstantiatorsRemovalListener implements
+    RemovalListener<ArtifactDescriptor, Instantiators> {
+
+    @Override
+    public void onRemoval(RemovalNotification<ArtifactDescriptor, Instantiators> notification) {
+      Closeables.closeQuietly(notification.getValue());
```

Comment:
sorry, just seeing this. the cache key is parent artifact descriptor, so the size of the cache (number of elements) would be less, though the size of the value could be bigger as more plugin methods are invoked. maybe we can set maximumWeight with a custom weigher.